util/ipc_sync.c

```diff
@@ -0,0 +1,238 @@
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "noeintr.h"
+#include "warnp.h"
+
+#include "ipc_sync.h"
+
+/* Convention for read/write ends of a pipe. */
+#define R 0
+#define W 1
+
+struct ipc_sync {
+	int fd[2];
+};
+
+/* Read and discard one byte from ${fd}, looping upon EINTR. */
+static inline int
+readbyte(int fd)
+{
+	char dummy;
+	char done = 0;
+
+	/* Loop until done. */
+	do {
+		switch (read(fd, &dummy, 1)) {
+		case -1:
+			/* Anything other than EINTR is bad. */
+			if (errno != EINTR) {
+				warnp("read");
+				goto err0;
+			}
+
+			/* Otherwise, loop and read again. */
+			break;
+		case 0:
+			warn0("Unexpected EOF in pipe");
+			goto err0;
+		case 1:
+			/* Expected value; quit the loop. */
+			done = 1;
+		}
+	} while (!done);
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Write one byte from ${fd}, looping upon EINTR. */
+static inline int
+writebyte(int fd)
+{
+	char dummy = 0;
+
+	/* Do the write. */
+	if (noeintr_write(fd, &dummy, 1) == -1) {
+		warnp("write");
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* close(), but looping upon EINTR. */
+static inline int
+ipc_sync_close(int fd)
```

Comment:
Two notable things about this function:
1) we will replace it once we have `noeintr_close()`.
2) this function retains the previous behaviour of `daemonize.c`, namely that if `close()` has `EINTR`, this function calls `close()` again.  This behaviour is required for HP-UX, but is explicitly wrong for the glibc implementation of `close()`: https://www.man7.org/linux/man-pages/man2/close.2.html

I propose that we keep the same behaviour as before (which is broken on linux), but I thought that I should bring it to your attention.